docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,113 @@
+* See **Hook Updates** section for more details on these hooks.
+
+### Generator Config Updates
+
+A new `LateInitializationConfig` struct will be introduced which will become a member of `ResourceConfig` struct.
+
+`LateInitializationConfig` struct will look like
+
+```
+type LateInitializationConfig struct { 
+   
+   // valid values: 'readOne, create, update'
+   DefaultSourceMethod *string `json:"default_source_method"`
+   
+   LateInitializedFields []LateInitializedFieldConfig `json:"late_initialized_fields,omitempty"`
+}
+
+
+type LateInitializedFieldConfig struct {
+   
+   // valid values: 'readOne, create, update'
+   // If this is not mentioned, then default_source_method is used
+   SourceMethod *string `json:"source_method,omitempty"`
+   
+   // Path to the field relative to Spec, which will be late initialized
+   // Example for primitives: "a" to replace "desired.ko.Spec.a" with "<source_method>Latest.ko.spec.a"
+   // Example for struct: "a.b" to replace "desired.ko.Spec.a.b" with "<op_name>Latest.ko.spec.a.b"
+   // Example for map: "a.b.[c]" to replace "desired.ko.Spec.a.b[c]" with "<op_name>Latest.ko.spec.a.b[c]"
+   // TODO: support for List
+   Path string `json:"path"`   
+}
+```
+
+### Hook Updates
+
+There will be 3 new hooks added for dealing with late initialized fields
+
+1. `add_late_initialized_fields_override` : The template used in this hook will be used as exact implementation of ResourceDescriptor’s `AddLateInitializedFields()` method
+
+2. `pre_add_late_initialized_fields`: The template used in this hook will be inserted in beginning of the ResourceDescriptor’s `AddLateInitializedFields` method. The variables available to be used in the template will be `desired`, `readOneLatest`, `createLatest`, `updateLatest`, `specUpdated`(equal to false) and `err`(equal to nil)
+
+3. `post_add_late_initialized_fields`: The template used in this hook will be inserted before the last `return specUpdated, err` statement of ResourceDescriptor’s `AddLateInitializedFields()` method. The variables available to be used in the template will be `desired`, `readOneLatest`, `createLatest`, `updateLatest`, `specUpdated` and `err`.
+
+
+Based on the Requirements above, this mentioned solution will: 
+
+1. Update the desired state of the K8s custom resource with late initialized fields.  ✅
+
+2. Provide configuration for AWS service teams to indicate which operation output returns the late initialized fields. i.e. Get/List/Create/Update output. ✅
+
+3. Wait for asynchronous resource creation. ✅
+
+    > Service teams can add a hook which will look at the state of latest object and determine if the reconciliation loop should run again. If so, they can return an error from the “AddLateInitializedFields()” method and the reconciliation loop will be triggered again.
+
+4. Allow the service teams to provide custom code using hooks for handing of late initialized fields. ✅
```

Comment:
Did you evaluate other possible solutions? If not, lets at least add the discussions from this Github PR into the document for posterity.